### DISCUSSION TITLE
How to obtain the server data under PRESENCE_UPDATE event?

### DISCUSSION THREAD

**Discussion by pkSourceCode** [discord-php/DiscordPHP#1171] [Q&A] [answered]
**Hello,** 

Could you please if this code is correct or not?  I was trying to get the data out of the server, but when I print_r it doesn't show all of the server information. Thank you

```
$guild_id = $presence->guild_id;
        echo PHP_EOL . PHP_EOL;
        $discord->guilds->freshen()->done(function (GuildRepository $guilds) use ($guild_id)
        {
            echo "Freshen " . PHP_EOL;
            echo "\t Now loading data of Guild: " . $guild_id . PHP_EOL;
        });
        $discord->guilds->fetch($guild_id)->done(function (Guild $guild)
        {
            print_r(json_encode($guild), JSON_PRETTY_PRINT);
        });

```

I've tried
```
$guild_id = $presence->guild_id;
        echo PHP_EOL . PHP_EOL;
        $discord->guilds->freshen()->done(function (GuildRepository $guilds) use ($guild_id)
        {
            echo "Freshen ...." . PHP_EOL;
            echo "\t Now loading data of Guild: " . $guild_id . PHP_EOL;
            echo "How Many Servers?: " . $guilds->count() . PHP_EOL;
            $guilds->fetch($guild_id)->done(function (Guild $guild)
            {
                echo "Member Count:" . $guild->members->count() . PHP_EOL;
            });
        });
```
The members contains nothing in it. Please help

Thank you :)

**Comment by pkSourceCode** [comment]
```
$discord = new Discord([
    'token' => getKey(),
    'loadAllMembers' => true,
    'intents' => [Intents::GUILD_PRESENCES,
                  Intents::GUILD_MEMBERS,
                  Intents::GUILD_MEMBERS,
                  Intents::GUILD_MESSAGES]
]);
```

**Comment by PeanutNL** [comment]
You can use `$presence->guild` to get the guild data

**Reply by pkSourceCode** [reply]
Hello sir. thank you for the comment.

but it seems like $presence->guild->member is also NULL

I tried
` print_r(json_encode($presence->guild, JSON_PRETTY_PRINT));`

it prints out NULL in the console

The bot has admin permission. I don't know if there's anything possibly wrong with the set up?

Thank you

**Reply by PeanutNL** [reply]
its `$presence->guild->members`
check: https://github.com/discord-php/DiscordPHP/blob/b34811b648e6c121f54769cfdcf7fd75d8c26e7a/src/Discord/Parts/Guild/Guild.php#L137

**Reply by pkSourceCode** [reply]
Thank you for your help!!!
I tried print_r the $presence and here is the response
```
{
    "user": {
        "id": "myuserid",
        "username": null,
        "discriminator": null,
        "avatar": "https:\/\/cdn.discordapp.com\/embed\/avatars\/0.png?s
ize=1024",
        "bot": null,
        "system": null,
        "mfa_enabled": null,
        "locale": null,
        "verified": null,
        "email": null,
        "flags": null,
        "banner": null,
        "accent_color": null,
        "premium_type": null,
        "public_flags": null
    },
    "guild_id": "1142028390279417897",
    "status": "idle",
    "activities": [],
    "client_status": {
        "web": "idle"
    },
    "guild": null,
    "game": null,
    "desktop_status": null,
    "mobile_status": null,
    "web_status": "idle",
    "member": null,
    "roles": []
}
```
From your solution.
`print_r(json_encode($presence->guild->members, JSON_PRETTY_PRINT));`

Warning: Attempt to read property "members" on null in H:\php8\discord-b
ots\bot.php on line 93
null


Is there any additional headings to add when building the Discord instance? 

Please help Thank you!!

**Reply by pkSourceCode** [reply]
Why are all of these NULL? Please help

Thank you very much

**Reply by pkSourceCode** [reply]
I installed DiscordPHP following these guidelines

https://github.com/discord-php/DiscordPHP/tree/b34811b648e6c121f54769cfdcf7fd75d8c26e7a#discordphp

I'll try reinstall it again soon.

**Reply by PeanutNL** [reply]
According to https://discord.com/developers/docs/topics/gateway-events#presence-update The user object can be partial.
And according to https://github.com/discord-php/DiscordPHP/blob/b34811b648e6c121f54769cfdcf7fd75d8c26e7a/src/Discord/Parts/WebSockets/PresenceUpdate.php#L33 The guild can be NULL so you have to get the guild from the cache.
Try:
`if ($presence->guild_id !== NULL)
{
       $discord->guilds->cache->get($presence->guild_id)->then(function(Guild $guild)
        {
            print_r(json_encode($guild), JSON_PRETTY_PRINT);
            echo "Member Count:" . $guild->members->count() . PHP_EOL;
        });
}`

**Reply by pkSourceCode** [reply]
Thank you very much for your time

I think there's something wrong with my php or DiscordPHP installation.
My PHPStorm could not find the  $discord->guilds->cache,

After running 

_Fatal error: Uncaught Error: Call to a member function get() on null_ 

I'm using terminal on PHPStorm and when I use the window CMD to run PHP. It gave me a __DiscordPHP.ERROR: failed to connect to websocket, retry in 5 seconds__ 

This goes forever. So I think that might be a problem to fix.

I'll try figure it out, and will update the situation soon.

Thank you very much!

**Comment by pkSourceCode** [comment, marked as answer]
It solved now!!!

The problem is in the Discord instance declaration. I add intents with array object. As soon as I changed to | like
```
 'intents' =>  Intents::GUILD_MEMBERS |
                      Intents::GUILD_MESSAGES |
                      Intents::getDefaultIntents()
```

Everything works. The problem wasn't about network or installation or anything, but about the **intent in the Discord instance declaration** above 
'
Separate the Intents with the | sign, not with the array [Intent1, Intent2, Intent3, ....].
